designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,326 @@
+  // Initially, this will be marked as alpha (v1alpha1).
+  APIVersion    string              `json:"apiVersion"`
+
+  // Args holds the plugin specific arguments that are received from the CLI which are to be passed down to the plugin.
+  Args          []string            `json:"args"`
+
+  // Universe represents the modified file contents that gets updated over a series of plugin runs
+  // across the plugin chain. Initially, it starts out as empty.
+  Universe      map[string]string   `json:"universe"`
+}
+
+// PluginResponse is returned to kubebuilder by the plugin and contains all files
+// written by the plugin following a certain command.
+type PluginResponse struct {
+  // Command holds the command that gets executed by the plugin such as init, create api, etc.
+  Command       string                   `json:"command"`
+
+  // Help contains the plugin specific help text that the plugin returns to kubebuilder when it receives
+  // `--help` flag from Kubebuilder.
+  Help          string                   `json:"help,omitempty"`
+
+  // APIVersion defines the versioned schema of the PluginResponse that will be written back to kubebuilder.
+  // Initially, this will be marked as alpha (v1alpha1).
+  APIVersion    string                   `json:"apiVersion"`
+
+  // Universe in the PluginResponse represents the updated file contents that was written by the plugin.
+  Universe      map[string]string        `json:"universe"`
+
+  // Error is a boolean type that indicates whether there were any errors due to plugin failures.
+  Error         bool                     `json:"error,omitempty"`
+
+  // ErrorMsg holds the specific error message of plugin failures.
+  ErrorMsg      string                   `json:"error_msg,omitempty"`
+}
+```
+
+These structs would be defined on the Kubebuilder side. `PluginRequest` holds all the information `kubebuilder` receives from the CLI and the request will be marshaled into a JSON and sent over `stdin`. `PluginResponse` is what the plugin constructs with the updated universe and sent back to `kubebuilder`.
+
+```go
+func (p *ExternalPlugin) runExternalProgram(req PluginRequest) (res PluginResponse, err error) {
+  pluginReq, err := json.Marshal(req)
+  if err != nil {
+    return res, err
+  }
+
+  cmd := exec.Command(p.Path)
+  cmd.Dir = p.DirContext
+  cmd.Stdin = bytes.NewBuffer(pluginReq)
+  cmd.Stderr = os.Stderr
+
+  out, err := cmd.Output()
+  if err != nil {
+    fmt.Fprint(os.Stdout, string(out))
+    return res, err
+  }
+
+  if json.Unmarshal(out, &res); err != nil {
+    return res, err
+  }
+  return res, nil
+}
+```
+
+The above function handles construction of the `PluginRequest` based on the information `kubebuilder` receives from the CLI and the request is marshaled into JSON. The command to run the external plugin by providing the plugin path will be invoked and `kubebuilder` will send the marshaled `PluginRequest` JSON to the plugin over `stdin`.
```

Comment:
Move this above the `runExternalPlugin` code and switch it to "The following function handles ..." It's easier to read IMO to have a slight intro before code dumps.